### DISCUSSION TITLE
Snapshots tests fail when Json returns in different order,

### DISCUSSION THREAD

**Discussion by jambudi** [pointfreeco/swift-snapshot-testing#807] [General]
How do we resolve the problem when JSON returns in different orders all the time?

![Screenshot 2023-11-08 at 4 07 38 PM](https://github.com/pointfreeco/swift-snapshot-testing/assets/25522050/08b77a36-63c4-46dd-a4f0-bda6ca4c9611)

**Comment by mbrandonw** [comment]
Hi @jambudi, the `.curl` snapshot strategy does not have any awareness of JSON bodies vs HTTP form bodies vs any other kind of body. All it sees is [raw data that is encoded to a string](https://github.com/pointfreeco/swift-snapshot-testing/blob/bb0ea08db8e73324fe6c3727f755ca41a23ff2f4/Sources/SnapshotTesting/Snapshotting/URLRequest.swift#L97-L105).

As such there is nothing the strategy can do to stabilize the data format. Instead, the code producing the JSON string (somewhere in your code) should be responsible for outputting sorted JSON during tests.

As this is not an issue with the library I am going to convert this to a discussion.

**Reply by willmorris44** [reply]
@mbrandonw We ended up solving this by replacing the data in the URLRequest httpBody with sorted JSON data using `JSONSerialization.data(withJSONObject: jsonObject, options: [.sortedKeys])`. This seems to be an issue that others have run into. I think it could be very beneficial if the library tried turning the httpBody into a JSON object and sort the keys. If it fails, it should continue with the standard process. It would look something like this (with more comprehensive handling of the try's): 

```
if let httpBodyData = request.httpBody {
    if let jsonObject = try? JSONSerialization.jsonObject(with: httpBodyData) as? [String: Any],
        let jsonData = try? JSONSerialization.data(withJSONObject: jsonObject, options: [.sortedKeys]),
        let httpBody = String(data: jsonData, encoding: .utf8) {
           // Parse and append here 
    } else {
        // Do normal process here
    }
}
```

This will sort the keys of the JSON dictionary alphabetically and will keep the tests stable and consistent for JSON bodies. 

Since API's don't typically require keys to be in any particular order, but the library does, I think it would be helpful for new consumers of the library to not have to worry about sorting the JSON themselves in their tests.
